contracts/core/Allo.sol

```diff
@@ -1,79 +1,359 @@
+    /// @param _identityId The identityId of the pool creator in the registry
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token that the pool is denominated in
+    /// @param _amount The amount of the token to be deposited into the pool
+    /// @param _metadata The metadata of the pool
+    function createPool(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata
+    ) external payable returns (uint256 poolId) {
+        return _createPool(_identityId, _allocationStrategy, _distributionStrategy, _token, _amount, _metadata);
     }
 
-    // @todo insert clonable strategy library, including validation that an existing pool has safe strategy
+    /// @notice Creates a new pool
+    /// @param _identityId The identityId of the pool creator in the registry
+    /// @param _allocationStrategy The address of the allocation strategy
+    /// @param _distributionStrategy The address of the distribution strategy
+    /// @param _token The address of the token that the pool is denominated in
+    /// @param _amount The amount of the token to be deposited into the pool
+    /// @param _metadata The metadata of the pool
+    function _createPool(
+        bytes32 _identityId,
+        address _allocationStrategy,
+        address payable _distributionStrategy,
+        address _token,
+        uint256 _amount,
+        Metadata memory _metadata
+    ) internal returns (uint256 poolId) {
+        if (!registry.isOwnerOrMemberOfIdentity(_identityId, msg.sender)) {
+            revert NO_ACCESS_TO_ROLE();
+        }
+
+        if (usedStrategies[_allocationStrategy] || usedStrategies[_distributionStrategy]) {
+            revert STRATEGY_ALREADY_USED();
+        }
+
+        usedStrategies[_allocationStrategy] = true;
+        usedStrategies[_distributionStrategy] = true;
+
+        Pool memory pool = Pool({
+            identityId: _identityId,
+            allocationStrategy: IAllocationStrategy(_allocationStrategy),
+            distributionStrategy: IDistributionStrategy(_distributionStrategy),
+            metadata: _metadata
+        });
+
+        poolId = _poolIndex++;
 
-    // creates pool locally, transfers pool amount to distribution strategy => returns poolId
-    // takes fee from user
-    // validates that the owner is actually allowed to use the identity
-    function createPool(PoolData memory /*_poolData*/ ) external pure returns (uint256) {
-        uint32 _poolId = 0;
+        if (_amount > 0) {
+            _fundPool(_token, _amount, poolId, address(pool.distributionStrategy));
+        }
 
-        // todo: return the poolId? what do we want to return here?
-        return _poolId;
+        pools[poolId] = pool;
```

Comment:
I haven't thought through this completely but this doesn't follow `checks-effects-interactions` because `_fundPool` is reenterable (external call of ETH to untrusted address). 

Could lead to reentrancy risk where someone creates a pool, then calls back to themselves and can create another pool they shouldn't be allowed to because `pools[poolId]` hasn't been set with that info yet?

I'd flip the order of these to be safe.

Reply:
Good call-out @zobront 

For our reference:
The CEI pattern prevents reentrancy attacks by ensuring that all state changes are made before any interactions with other contracts occur. This is done by dividing the code of a function into three phases:

Checks: This phase performs all necessary checks on the input parameters and the current state of the contract.
Effects: This phase makes any changes to the state of the contract.
Interactions: This phase makes calls to other contracts.

By following this pattern, it is guaranteed that any state changes will be committed before any interactions with other contracts occur. This prevents malicious contracts from calling a function in another contract and then calling that function again before the first call has finished executing.

Reply:
Looks like this is taken care of. Just checking

Reply:
Updating.

```javascript
poolId = ++_poolIndex;
pools[poolId] = pool;

if (_amount > 0) {
  _fundPool(_token, _amount, poolId, address(pool.distributionStrategy));
  emit PoolCreated(poolId, _identityId, _allocationStrategy, _distributionStrategy, _token, _amount, _metadata);
}

// forcing 0 for the amount bc that either means the tx to transfer failed or they didn't send any funds
emit PoolCreated(poolId, _identityId, _allocationStrategy, _distributionStrategy, _token, 0, _metadata);
```